File: src/main/java/rusha/x/Spiral.java
package rusha.x;

import java.util.Scanner;

public class Spiral {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Integer m = scanner.nextInt();
        int[][] matrix = new int[m][];
        for (int i = 0; i < m; i++) {
            matrix[i] = new int[m];
            for (int j = 0; j < m; j++) {
                matrix[i][j] = scanner.nextInt();
            }
        }
        Direction direction = Direction.UP;

        int i = m / 2;
        int j = m / 2;
        int step = 1;
        step:
        while (true) {
            for (int a = 0; a < step; a++) {
                if (!(0 <= i && i < m && 0 <= j && j < m)) {
                    break step;
                }
                System.out.println(matrix[i][j]);
                switch (direction) {
                    case UP:
                        i -= 1;
                        break;
                    case RIGHT:
                        j += 1;
                        break;
                    case DOWN:
                        i += 1;
                        break;
                    case LEFT:
                        j -= 1;
                        break;
                }
            }
            switch (direction){

                case UP:
                    direction = Direction.RIGHT;
                    break;
                case RIGHT:
                    direction = Direction.DOWN;
                    step += 1;
                    break;
                case DOWN:
                    direction = Direction.LEFT;
                    break;
                case LEFT:
                    direction = Direction.UP;
                    step += 1;
                    break;
            }
        }

    }

    enum Direction {
        UP, RIGHT, DOWN, LEFT;
    }
}

File: src/main/kotlin/rusha/x/Spiral.kt
package rusha.x

import java.util.*

fun main() {
    val scanner = Scanner(System.`in`)
    val m = scanner.nextInt()
    val matrix = Array(m) {
        Array(m) {
            scanner.nextInt()
        }
    }

    var i = m / 2
    var j = m / 2
    var step = 1
    var direction = Direction.Up
    step@while (true) {
        for (a in 1..step) {
            if (!(0 <= i && i < m && 0 <= j && j < m)) {
                break@step
            }
            println(matrix[i][j])
            when (direction) {
                Direction.Up -> i -= 1
                Direction.Right -> j += 1
                Direction.Down -> i += 1
                Direction.Left -> j -= 1
            }
        }
        when (direction) {
            Direction.Up -> {
                direction = Direction.Right
            }
            Direction.Right -> {
                direction = Direction.Down
                step += 1
            }
            Direction.Down -> {
                direction = Direction.Left
            }
            Direction.Left -> {
                direction = Direction.Up
                step += 1
            }
        }
    }
}

enum class Direction {
    Up, Right, Down, Left
}